frame/treasury/src/lib.rs

```diff
@@ -501,6 +685,210 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// Propose and approve a spend of treasury funds.
+		///
+		/// ## Dispatch Origin
+		///
+		/// Must be [`Config::SpendOrigin`] with the `Success` value being at least
+		/// `amount` of `asset_kind` in the native asset. The amount of `asset_kind` is converted
+		/// for assertion using the [`Config::BalanceConverter`].
+		///
+		/// ## Details
+		///
+		/// Create an approved spend for transferring a specific `amount` of `asset_kind` to a
+		/// designated beneficiary. The spend must be claimed using the `payout` dispatchable within
+		/// the [`Config::PayoutPeriod`].
+		///
+		/// ### Parameters
+		/// - `asset_kind`: An indicator of the specific asset class to be spent.
+		/// - `amount`: The amount to be transferred from the treasury to the `beneficiary`.
+		/// - `beneficiary`: The beneficiary of the spend.
+		/// - `valid_from`: The block number from which the spend can be claimed. If `None`, the
+		///   spend can be claimed immediately after approval.
+		///
+		/// ## Events
+		///
+		/// Emits [`Event::AssetSpendApproved`] if successful.
+		#[pallet::call_index(5)]
+		#[pallet::weight(T::WeightInfo::spend())]
+		pub fn spend(
+			origin: OriginFor<T>,
+			asset_kind: T::AssetKind,
+			#[pallet::compact] amount: AssetBalanceOf<T, I>,
+			beneficiary: BeneficiaryLookupOf<T, I>,
+			valid_from: Option<T::BlockNumber>,
+		) -> DispatchResult {
+			let max_amount = T::SpendOrigin::ensure_origin(origin)?;
+			let beneficiary = T::BeneficiaryLookup::lookup(beneficiary)?;
+
+			let native_amount = T::BalanceConverter::from_asset_balance(amount, asset_kind.clone())
+				.map_err(|_| Error::<T, I>::FailedToConvertBalance)?;
+
+			ensure!(native_amount <= max_amount, Error::<T, I>::InsufficientPermission);
+
+			with_context::<SpendContext<BalanceOf<T, I>>, _>(|v| {
+				let context = v.or_default();
+				// We group based on `max_amount`, to distinguish between different kind of
+				// origins. (assumes that all origins have different `max_amount`)
+				//
+				// Worst case is that we reject some "valid" request.
+				let spend = context.spend_in_context.entry(max_amount).or_default();
+
+				// Ensure that we don't overflow nor use more than `max_amount`
+				if spend.checked_add(&native_amount).map(|s| s > max_amount).unwrap_or(true) {
+					Err(Error::<T, I>::InsufficientPermission)
+				} else {
+					*spend = spend.saturating_add(native_amount);
+					Ok(())
+				}
+			})
+			.unwrap_or(Ok(()))?;
+
+			let index = SpendCount::<T, I>::get();
+			let valid_from = valid_from.unwrap_or(frame_system::Pallet::<T>::block_number());
```

Comment:
```suggestion
			let now = frame_system::Pallet::<T>::block_number();
			let valid_from = if let Some(v) = valid_from {
				// Can't make a spend valid from the past.
				v.max(now)
			} else {
				now
			};
```

Reply:
I made it possible to point to the past, to not fail a proposal, which might contain multiple spend approves.
A user could be wrong for a couple of blocks or it took longer to approve a proposal.

Setting valid_from to now in such a case, is a different approach. A final spend might be different from the one a stakeholder left a vote for.
What you think?

Reply:
That is a good point about inadvertently pushing one call into a different Spend Period than intended. I was thinking of the [schedule enactment](https://github.com/paritytech/substrate/blob/b72c475/frame/referenda/src/lib.rs#L861) logic, with the idea that if a proposal passes, we don't want it to fail because the actual voting phase took longer than expected (maybe it took some time to secure a decision deposit).

I think you may be right, but we do need something to do if `valid_from + SpendPeriod < now`, because now something will have been approved but never actually claimable.

Reply:
updated